src/tests/networking_tests.rs

```diff
@@ -2421,9 +2421,348 @@ pub mod net_tests {
+
+            assert_eq!(cage2.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        });
+
+        let mut sockgarbage = interface::GenSockaddr::V4(interface::SockaddrV4::default());
+        let fd = cage.accept_syscall(serverfd, &mut sockgarbage);
+        assert!(fd > 0);
+
+        // first make sure send and recv are working
+        let mut buf = sizecbuf(11);
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 11), 11);
+        assert_eq!(cbuf2str(&buf), "client send");
+        assert_eq!(cage.send_syscall(fd, str2cbuf("server send"), 11, 0), 11);
+
+        // peer SHUT_WR
+        barrier.wait();
+        let mut buf = sizecbuf(14);
+        // data already sent should still be readable
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 14), 14);
+        assert_eq!(cbuf2str(&buf), "before SHUT_WR");
+        // subsequent read should return end-of-file
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 14), 0);
+
+        assert_eq!(cage.close_syscall(fd), 0);
+
+        let barrier2 = Arc::new(Barrier::new(2));
+        let barrier2_clone = barrier2.clone();
+
+        let thread2 = interface::helper_thread(move || {
+            let cage3 = interface::cagetable_getref(3);
+
+            let fd = cage3.socket_syscall(AF_INET, SOCK_STREAM, 0);
+            assert_eq!(cage3.connect_syscall(fd, &socket), 0);
+
+            // first make sure send and recv are working before shutdown
+            assert_eq!(cage3.send_syscall(fd, str2cbuf("client send"), 11, 0), 11);
+            let mut buf = sizecbuf(11);
+            assert_eq!(cage3.read_syscall(fd, buf.as_mut_ptr(), 11), 11);
+            assert_eq!(cbuf2str(&buf), "server send");
+
+            assert_eq!(
+                cage3.send_syscall(fd, str2cbuf("before SHUT_RDWR"), 16, 0),
+                16
+            );
+            assert_eq!(cage3.netshutdown_syscall(fd, SHUT_RDWR), 0);
+
+            barrier2_clone.wait();
+
+            assert_eq!(cage3.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        });
+
+        let fd = cage.accept_syscall(serverfd, &mut sockgarbage);
+        assert!(fd > 0);
+
+        // first make sure send and recv are working
+        let mut buf = sizecbuf(11);
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 11), 11);
+        assert_eq!(cbuf2str(&buf), "client send");
+        assert_eq!(cage.send_syscall(fd, str2cbuf("server send"), 11, 0), 11);
+
+        // peer SHUT_RDWR
+        barrier2.wait();
+        let mut buf = sizecbuf(16);
+        // data already sent should still be readable
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 16), 16);
+        assert_eq!(cbuf2str(&buf), "before SHUT_RDWR");
+        // subsequent read should return end-of-file
+        assert_eq!(cage.read_syscall(fd, buf.as_mut_ptr(), 16), 0);
+
+        thread.join().unwrap();
+        thread2.join().unwrap();
+
+        assert_eq!(cage.exit_syscall(EXIT_SUCCESS), EXIT_SUCCESS);
+        lindrustfinalize();
+    }
```

Comment:
Just wondering, did you find the bugs mentioned above in netshutdown_unix by comparing the errors returned there to the errors returned in netshutdown_inet?

Reply:
All these behaviors are specified in standards so I am comparing them with it. I'd recommand you to read The Linux Programming Interface by Michael Kerrisk (who is also the author of man page website), it may mention a lot more details about the syscall than man page